design-proposals/storage-migration/storage-migration.md

```diff
@@ -0,0 +1,282 @@
+   volumes. Hence, the API covers and refers only to the storage migration.
+
+ ## Definition of Users
+ This feature addresses for all KubeVirt users who want to move to another storage classes or PVCs without workload interruptions.
+
+## User Stories
+  1. A storage class of volumes used by a running VM is deprecated
+  2. A more performant storage class is now available in the cluster and the user wants start using it.
+  3. The storage device is filling up
+
+## Repos
+
+  - [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+This proposal introduces a new CRD `StorageMigration`
+
+```golang
+type StorageMigration struct {
+        metav1.TypeMeta   `json:",inline"`
+        metav1.ObjectMeta `json:"metadata,omitempty"`
+        Spec StorageMigrationSpec `json:"spec"`
+        Status *StorageMigrationStatus `json:"status,omitempty"`
+}
+
+// StorageMigrationList is a list of StorageMigration resources
+type StorageMigrationList struct {
+        metav1.TypeMeta `json:",inline"`
+        metav1.ListMeta `json:"metadata"`
+        Items []StorageMigration `json:"items"`
+}
+
+type MigratedVolume struct {
+        SourcePvc      string `json:"sourcePvc" valid:"required"`
+        DestinationPvc string `json:"destinationPvc" valid:"required"`
+        // ReclaimPolicySourcePvc describes how the source volumes will be
+        // treated after a successful migration
+        ReclaimPolicySourcePvc ReclaimPolicySourcePvc `json:"reclaimPolicySourcePvc,omitempty"
+}
+
+type MigrationStorageClass struct {
+        SourceStorageClass      string `json:"sourceStorageClass" valid:"required"`
+        DestinationStorageClass string `json:"destinationStorageClass" valid:"required"`
+        // PrefixDestinationVolume is the prefix used by KubeVirt to generate
+        // the destination PVC. The claim name will contain the prefix plus a
+        // random string.
+        PrefixDestinationVolume string `json:"PrefixDestinationVolume" valid:"required"`
+        // ReclaimPolicySourcePvc describes how the source volumes will be
+        // treated after a successful migration
+        ReclaimPolicySourcePvc ReclaimPolicySourcePvc `json:"reclaimPolicySourcePvc,omitempty"`
+}
+
+// ReclaimPolicySourcePvc describes how the source PVC will be treated after the storage migration completes. The policies follows the same behavior as the RetainPolicy for PVs: https://kubernetes.io/docs/concepts/storage/persistent-volumes/#reclaiming
+type ReclaimPolicySourcePvc string
+
+const (
+        DeleteReclaimPolicySourcePvc ReclaimPolicySourcePvc = "Delete"
+        RetainReclaimPolicySourcePvc ReclaimPolicySourcePvc = "Retain"
+)
+
+// StorageMigrationSpec is the spec for a StorageMigration resource
+type StorageMigrationSpec struct {
+        // The name of the VMI to perform the migration on. VMI must exist in the storage migration objects
+        // namespace
+        VMIName string `json:"vmiName,omitempty" valid:"required"`
```

Comment:
@mhenriks those are all valuable feedback!
However, this is a quite huge feature. We could add this incrementally and make it the more and more configurable. 
Additionally, we could also handle stopped VM in a second moment? I can change `VMIName` to `VMName`. However, we could have cases when we only have VMIs and not VM. Do you think we need to have 2 separate fields?
Right now, VMI always have the same name as the VM, but not sure if this will be the same in the future.

Reply:
@alicefr Just to confirm.  My main point is that there should not be a VMIName/VMName field.  Migrations should not be limited to a single VM/VMI.  But you want to limit to a single VM/VMI for now?

Reply:
Ah, I see then I misread this. The idea here was to create a `StorageMigration` object per VMI/VM. You want something more general based on storage classes? Then, the proposal will look quite different.

Reply:
> You want something more general based on storage classes?

Yeah, thinking `MigrationStorageClasses` would specify source/target storageClass and labelSelectors that would apply to a set of VM/VMIs

`MigrationVolumes` would be used to migrate a set of PVCs.  When the migration runs, the affected VM/VMIs would be looked up.  We already have indexes/informers for that mapping.

Reply:
The problem with a single object for multiple VMs is what if some migrations fail and others not. How do you report it? Additionally, as a user, I might not want to migrate all the VMs that use a certain storage class.

However, I get the idea that this CRD only uses storage information and your suggestion makes a lot of sense.  We could solve the failures somehow. This is about usability, I'm not sure what could be more user-friendly.

@vasiliy-ul WDYT?

Reply:
To me, it seems like a lot of extra work for the user to find all affected VMs using a PVC in a particular storage class.  And by using label selectors, users can make sure to limit to a subset of VMs/VMIs.

Reply:
@fabiand the storage class method is indeed more generic. However, if the user wants to have a more fine-grained control of the volumes, then it can always choose to list all the volumes to migrate

Reply:
However, if this CR triggers too many migrations at the same time, this could considerably slow down the cluster performance. I think we would need to define some way on the admin side to configure the migrations sequence. For example, the possibility to migrate the VM sequentially

Reply:
In a second version of this proposal, I'm planning to get rid of the VMI name